### PR TITLE
🔀 Bad response handler.

### PR DESCRIPTION
**Pull request by mom1**
Flexible conversion of an error into a clear client exception.

I think  this will be useful. 
I need help writing documentation.

### CODE REVIEW

**Comment by MikeWooster** [comment]
thanks for putting this together.  I've been super busy recently, but haven't forgotten about this.  I will take a closer look and get back to you over the next month.

**Comment by MikeWooster** [comment]
Okay, so I think this is a nice idea.  I would push to add an `ExceptionHandler` to the main client class which has a factory method to retrieve the client exception class from the status code.  Something like:
```
class ExceptionHandler:
	def get_exception_class(status_code: int) -> Type[APIRequestError]:
		return ClientError
```

The request strategy could call this on the parent client and then continue to raise as is.  This would allow consumers of the client to alter the type of exception raised, although it will still need to conform to the `APIRequestError`.

An alternative would be to refactor the response into a plain dataclass to abstract away from `requests` response implementation, and pass that into the ExceptionHandler so that you can add a method to return an exception which the strategy could raise.  This would allow for greater control of the exception from the consumers POV, but would involve a lot more work.

**Comment by mom1** [comment]
I hope I understood you correctly. My English is not very good :)

If everything is ok, then I will collect everything in one commit

File: apiclient/exceptions.py [hunk 1]
```diff
@@ -18,10 +20,14 @@ class ResponseParseError(APIClientError):
 class APIRequestError(APIClientError):
     """Exception to denote that something went wrong when making the request."""
 
-    def __init__(self, message, status_code: OptionalInt = None, info: str = ""):
-        self.message = message
-        self.status_code = status_code
-        self.info = info
+    message: str = ""
+    status_code: OptionalInt = None
+    info: str = ""
```
Comment (MikeWooster) [inline apiclient/exceptions.py:25]: you can remove these definitions.  They aren't used.

File: apiclient/request_strategies.py [hunk 2]
```diff
@@ -134,24 +134,20 @@ def _decode_response_data(self, response: Response):
 
     def _handle_bad_response(self, response: Response):
         """Convert the error into an understandable client exception."""
-        exception_class = self._get_exception_class(response.status_code)
+        exception_class = self._get_exception_from_client(response)
+        if not exception_class:
+            exception_class = self._get_exception_class(response.status_code)
```
Comment (MikeWooster) [inline apiclient/request_strategies.py:139]: I would push to keep this as it was, and remove the exception class from client.

File: apiclient/request_strategies.py [hunk 3]
```diff
@@ -134,24 +134,20 @@ def _decode_response_data(self, response: Response):
...
-            exception_class = exceptions.RedirectionError
-        elif 400 <= status_code < 500:
-            exception_class = exceptions.ClientError
-        elif 500 <= status_code < 600:
-            exception_class = exceptions.ServerError
-        else:
-            exception_class = exceptions.UnexpectedError
-        return exception_class
+    def _get_exception_from_client(self, response):
+        return self.get_client().get_response_handler().get_exception_class(response)
```
Comment (MikeWooster) [inline apiclient/request_strategies.py:147]: Remove this method.  In order to keep this simple, just using the exception handler on the parent client would be good enough for a customisable use case.

File: apiclient/response_handlers.py [hunk 4]
```diff
@@ -23,6 +27,34 @@ class RequestsResponseHandler(BaseResponseHandler):
     def get_request_data(response: Response) -> Response:
         return response
 
+    @classmethod
+    def get_exception_class(cls, response: Response) -> Optional[Type[Exception]]:
+        return None
+
+
+class RequestsExceptionByKeyMixin:
```
Comment (MikeWooster) [inline apiclient/response_handlers.py:35]: This should be necessary if we remove everything except the exception handler on the main client.

**Review by MikeWooster — COMMENTED** [review]
To keep it simple, there should only be one way of handling exceptions, with the most logical place for it to live on the main client.  I've pointed out a few bits that should be removed to facilitate this.  Hope it makes sense.

File: apiclient/response_handlers.py (hunk 4 above)
Comment (mom1) [reply apiclient/response_handlers.py:35]: I didn't get it. Remove it?

File: apiclient/exceptions.py (hunk 1 above)
Comment (mom1) [reply apiclient/exceptions.py:25]: 👌

File: apiclient/request_strategies.py (hunk 2 above)
Comment (mom1) [reply apiclient/request_strategies.py:139]: 👌

File: apiclient/request_strategies.py (hunk 3 above)
Comment (mom1) [reply apiclient/request_strategies.py:147]: 👌

**Comment by mom1** [comment]
Okay. But I would push to pass as parameter the entire `response` in
`ExceptionHandler.get_exception_class` so that it would be possible to select an exception class based on the content of the response.

File: apiclient/exceptions.py (hunk 1 above)
Comment (mom1) [reply apiclient/exceptions.py:27]: This is needed for child classes. For example, to set values ​​that will override what is passed.
For example:
Server return in payload error code 20001 `{"error_code": 20001}`. This mean some error. I want seen clear message  in exception:
```python
class WrongCardNumber(APIRequestError):
    message = "Wrong card number"

ERROR_CODES_WITH_EXCEPTIONS_MAP = {
    20001: WrongCardNumber
}

class MyExceptionHandler(ExceptionHandler):
    @staticmethod
    def get_exception_class(response: Response) -> Type[APIRequestError]:
        return ERROR_CODES_WITH_EXCEPTIONS_MAP.get(response.json().get('error_code'))
```

File: apiclient/response_handlers.py (hunk 4 above)
Comment (mom1) [reply apiclient/response_handlers.py]: 👌

**Comment by MikeWooster** [comment]
> Okay. But I would push to pass as parameter the entire `response` in
> `ExceptionHandler.get_exception_class` so that it would be possible to select an exception class based on the content of the response.

I agree.  Although I think it would be worth introducing a response abstraction.

**Comment by mom1** [comment]
> > Okay. But I would push to pass as parameter the entire `response` in
> > `ExceptionHandler.get_exception_class` so that it would be possible to select an exception class based on the content of the response.
> 
> I agree. Although I think it would be worth introducing a response abstraction.

I don't quite understand. What do you mean?

**Comment by MikeWooster** [comment]
I've created a new PR (#57) that will better explain what I mean, and should allow you to modify this PR to allow you to pass around the response.

**Comment by mom1** [comment]
ok. I understood and will to modify this PR.

**Comment by MikeWooster** [comment]
> ok. I understood and will to modify this PR.

excellent.  I have merged the other PR so you have something to work from.

**Comment by mom1** [comment]
> > ok. I understood and will to modify this PR.
> 
> excellent. I have merged the other PR so you have something to work from.

Done

File: apiclient/exceptions.py [hunk 5]
```diff
@@ -59,3 +66,24 @@ class UnexpectedError(APIRequestError):
...
     pass
+
+
+class BaseExceptionHandler:
+    @staticmethod
+    def get_exception_class(response: Response) -> Type[APIRequestError]:
+        raise NotImplementedError
+
+
+class ExceptionHandler(BaseExceptionHandler):
```
Comment (MikeWooster) [inline apiclient/exceptions.py:77]: small request: could you rename this to `ErrorHandler`  as it's not actually handling an exception, but rather handling an error to return an exception type.  I think this should also be moved to a different module `error_handlers.py`, to keep the `exceptions.py` simply for defining exception types.

File: apiclient/exceptions.py (hunk 1 above)
Comment (MikeWooster) [reply apiclient/exceptions.py:26]: okay, I'm happy with this.

File: apiclient/exceptions.py [hunk 6]
```diff
@@ -59,3 +66,24 @@ class UnexpectedError(APIRequestError):
     """An unexpected error occurred when making the request."""
 
     pass
+
+
+class BaseExceptionHandler:
```
Comment (MikeWooster) [inline apiclient/exceptions.py:71]:
```suggestion
class BaseExceptionHandler:
    """Translates a response into an apiclient exception."""
```
Needs some docstrings.

**Review by MikeWooster — COMMENTED** [review]
Thanks for doing this.  I think this is almost there.  There's a few small changes that I've requested, and then we can add some documentation to the README.

File: apiclient/exceptions.py (hunk 5 above)
Comment (mom1) [reply apiclient/exceptions.py]: 👌

File: apiclient/exceptions.py (hunk 6 above)
Comment (mom1) [reply apiclient/exceptions.py]: 👌

**Comment by mom1** [comment]
I'm confused. 😄 
In CI black says `would reformat /home/runner/work/api-client/api-client/apiclient/error_handlers.py`
If I do this then isort will says `ERROR: /Users/maxst/projects/api-client/apiclient/error_handlers.py Imports are incorrectly sorted and/or formatted.`

**Comment by mom1** [comment]
> Thanks for doing this. I think this is almost there. There's a few small changes that I've requested, and then we can add some documentation to the README.

Done

**Review by MikeWooster — APPROVED** [review]

**Comment by MikeWooster** [comment]
I've been playing around with this PR a bit trying to write some docs.  I think we could put more of the burden in the error handler itself.  I think it should return the exception, rather than the exception class.  This means you can construct whatever exception you like, i.e. put in the message, or whatever, and the request strategy can then raise the exception.  I don't think it'll break the implementation too much, and it would be cleaner to be able to remove more of the error handling logic out of the strategy.

Something like:
```
class BaseErrorHandler:

    @staticmethod
    def get_exception(response: Response) -> APIRequestError:
		...
```

Thoughts?

**Comment by mom1** [comment]
hmm. Sounds cool.
This may solve some of my problems.

I will do this a little later.

**Comment by mom1** [comment]
Done

**Review by MikeWooster — DISMISSED** [review]
excellent.  this makes it a lot easier to use.

**Comment by MikeWooster** [comment]
Would it be possible to squash your commits.  We can then work on the docs.

**Comment by mom1** [comment]
can't squash merge

**Comment by MikeWooster** [comment]
> can't squash merge

Ok, I ran the following to remove those merge commits in your history:
```
git reset HEAD~3
git commit -m 'Bad response handler.
Flexible conversion of an error into a clear client exception.'
git fetch
git rebase origin/master
# fix up the conflicts
# run tests locally to make sure it's working still
# `git add` all fixed files
git rebase --continue
git push -f
```

In future, when you want to bring in changes, it's best to use the rebase method as you go. That way you don't get the Merge commits in the history.

**Comment by mom1** [comment]
@MikeWooster Thanks

**Review by MikeWooster — APPROVED** [review]
excellent, thank you!

**Pull request merged**